msdm/algorithms/lrtdp.py

```diff
@@ -50,16 +51,14 @@ def plan_on(self, mdp: MarkovDecisionProcess):
         res = self.res
         res.policy = AssignmentMap()
         res.Q = AssignmentMap()
-        for s in mdp.state_list:
+        for s in sum(self.res.trials, []):
+            if s in res.policy:
+                continue
             res.policy[s] = AssignmentMap([(self.policy(mdp, s), 1)])
             res.Q[s] = AssignmentMap()
-            for a in mdp.action_list:
+            for a in mdp.actions(s):
                 res.Q[s][a] = self.Q(mdp, s, a)
-        res.policy = TabularPolicy(
-            states=mdp.state_list,
-            actions=mdp.action_list,
-            policy_dict=res.policy
-        )
+        res.policy = PartialPolicy(res.policy)
```

Comment:
Per phone conversation, could entirely avoid allocating extra memory here by making a Policy class that is backed by a value function.